doc/fluid/design/data_pipeline/data_pipeline.md

```diff
@@ -0,0 +1,92 @@
+# Design Doc: Fluid Data Pipeline
+
+This document is about how Fluid training and inference programs read data.
+
+## Standard Data Format
+
+### Case 1: Data from Files
+
+Consider a Fluid training program, `resnet50.py`, needs to read data from disk:
+
+```bash
+cat data | python resnet50.py
```

Comment:
A note about pipe.

A pipe is a special file. We can even use `mkfifo` to create a named pipe, and it just acts like a real file.
Currently, Fluid support an operator named `open_files`. It can open multiple files and read them concurrently. We can create many named pipes and use many process to generate data.

```bash
mkfifo data1
mkfifo data2
./data_generator > data1 &  # make this process running at background
./data_generator > data2 &  # data1 and data2 will be generated concurrently.

#  in train.py
#  file_obj = fluid.open_files(['data1', 'data2'])
#  img, label = read(file_obj)

python train.py
```

Reply:
I agree it is a good idea to improve the data generation performance by reading from multiple data generators via pipes.

I also want to note that it is not a good design to have `OpenFiles`; instead, what we need are finer grained APIs:

1. OpenFile -- File, but not Files, and
1. CreateMultiplexReader

which provides the usage of

```c++
r = CreateMultiplexReader(
      OpenFile("/afs/resnet50/traininig-data-00000-of-00005"),
      OpenFile("/afs/resnet50/traininig-data-00003-of-00005"))
```

Reply:
@JiayiFeng  MultiplexReader is an alternative way to implement open_files. Please take a look.